### PR TITLE
assorted fixes for bundles

### PR DESCRIPTION
**Pull request by xeno-by**
This commit clearly establishes what bundles are. Previously it was
enough to just extend scala.reflect.macros.Macro, which created some
loopholes, but now scalac enforces that bundles:
  1) Are static (not necessarily top-level, but just static)
  2) Are traits (objects shouldn't be bundles anyway, and classes bring
     complications with their ctors which require special treatment in
     generated classes, so why support them if they don't bring anything
     new to the table?)
  3) Are monomorphic (again, this brings unnecessary complications wrt
     auxiliary code generation, so I don't see merit in supporting
     polymorphic bundles, whatever that a polymorphic bundle could mean)
  4) Don't provide concrete implementation for Macro.c (if they do then
     what is the point?)

Also fixes a couple issues. Firstly, bundles can now be nested as long
as they are static (which means that we can finally use bundles in REPL!).
Secondly, auxiliary code that's generated by the compiler alongside bundles
is now put into virtual compilation units with stable names, which solves
the infinite compilation problem in SBT 0.13.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym @gkossakowski @harrah

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala [hunk 1]
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
...
 
-        // synthesize the invoker, i.e. given a top-level `trait Foo extends Macro { def expand = ... } `
-        // create a top-level definition `class FooInvoker(val c: Context) extends Foo` in MACRO_INVOKER_PACKAGE
-        val invokerPid = gen.mkUnattributedRef(nme.MACRO_INVOKER_PACKAGE)
-        val invokerName = TypeName(bundleClass.fullName.split('.').map(_.capitalize).mkString("") + nme.MACRO_INVOKER_SUFFIX)
-        val invokerFullName = TypeName(s"$invokerPid.$invokerName")
-        val existingInvoker = rootMirror.getClassIfDefined(invokerFullName)
-        if (!currentRun.compiles(existingInvoker)) {
+        // synthesize the bundle, i.e. given a top-level `trait Foo extends Macro { def expand = ... } `
+        // create a top-level definition `class FooBundle(val c: Context) extends Foo` in MACRO_BUNDLE_PACKAGE
```
Comment (retronym) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: Is "top-level" in the comments still accurate? Or are things more general now ("static")?

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala [hunk 2]
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
...
+          currentRun.compileLate(bundleFileName, PackageDef(bundlePid, List(bundle)))
         }
 
         // synthesize the macro impl reference, which is going to look like:
-        // `new Foo$invoker(???).expand` plus the optional type arguments
-        val instanceOfInvoker = New(Select(invokerPid, invokerName), List(List(Select(scalaDot(nme.Predef), nme.???))))
-        gen.mkTypeApply(Select(instanceOfInvoker, methName), targs)
+        // `new FooBundle(???).macroName` plus the optional type arguments
+        val bundleInstance = New(Select(bundlePid, bundleName), List(List(Select(scalaDot(nme.Predef), nme.???))))
+        gen.mkTypeApply(Select(bundleInstance, methName), targs)
```
Comment (retronym) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: Maybe I'm missing a later call to `atPos`, but this tree looks unpositioned.

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala [hunk 3]
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
...
+          val bundle = atPos(bundleProto.pos)(ClassDef(NoMods, bundleName, bundleProto.typeParams.map(TypeDef(_)), bundleTemplate))
+          currentRun.compileLate(bundleFileName, PackageDef(bundlePid, List(bundle)))
         }
 
         // synthesize the macro impl reference, which is going to look like:
-        // `new Foo$invoker(???).expand` plus the optional type arguments
-        val instanceOfInvoker = New(Select(invokerPid, invokerName), List(List(Select(scalaDot(nme.Predef), nme.???))))
-        gen.mkTypeApply(Select(instanceOfInvoker, methName), targs)
+        // `new FooBundle(???).macroName` plus the optional type arguments
+        val bundleInstance = New(Select(bundlePid, bundleName), List(List(Select(scalaDot(nme.Predef), nme.???))))
```
Comment (retronym) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: It's not part of this patch, but we have `definitions.Predef_???` that could be used here.

File: src/reflect/scala/reflect/internal/StdNames.scala [hunk 4]
```diff
@@ -311,9 +311,8 @@ trait StdNames {
     val FAKE_LOCAL_THIS: NameType          = "this$"
     val LAZY_LOCAL: NameType               = "$lzy"
     val LAZY_SLOW_SUFFIX: NameType         = "$lzycompute"
-    val MACRO_INVOKER_PACKAGE: NameType    = "scala.reflect.macros.synthetic"
-    // TODO: if I use dollars in MACRO_INVOKER_SUFFIX, as in "$Invoker$", then Scala reflection fails to load implementations
-    val MACRO_INVOKER_SUFFIX: NameType     = "Invoker"
+    val MACRO_BUNDLE_PACKAGE: NameType     = "scala.reflect.macros.bundles"
+    val MACRO_BUNDLE_SUFFIX: NameType      = "Bundle"
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/StdNames.scala]: Perhaps `$Bundle` would make it a bit clearer that this is a compiler generated class.

File: src/reflect/scala/reflect/internal/StdNames.scala [hunk 5]
```diff
@@ -311,9 +311,8 @@ trait StdNames {
     val FAKE_LOCAL_THIS: NameType          = "this$"
     val LAZY_LOCAL: NameType               = "$lzy"
     val LAZY_SLOW_SUFFIX: NameType         = "$lzycompute"
-    val MACRO_INVOKER_PACKAGE: NameType    = "scala.reflect.macros.synthetic"
-    // TODO: if I use dollars in MACRO_INVOKER_SUFFIX, as in "$Invoker$", then Scala reflection fails to load implementations
-    val MACRO_INVOKER_SUFFIX: NameType     = "Invoker"
+    val MACRO_BUNDLE_PACKAGE: NameType     = "scala.reflect.macros.bundles"
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/StdNames.scala]: Why not in the user's package of the bundle trait? I'm not comfortable with the compiler generating third party code underneath a global namespace, let alone one underneath `scala._`. Is there potential for clashes when mixing JARs defining bundles?

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala [hunk 6]
```diff
@@ -41,30 +41,34 @@ trait Resolvers {
...
-        // create a top-level definition `class FooInvoker(val c: Context) extends Foo` in MACRO_INVOKER_PACKAGE
-        val invokerPid = gen.mkUnattributedRef(nme.MACRO_INVOKER_PACKAGE)
-        val invokerName = TypeName(bundleClass.fullName.split('.').map(_.capitalize).mkString("") + nme.MACRO_INVOKER_SUFFIX)
-        val invokerFullName = TypeName(s"$invokerPid.$invokerName")
-        val existingInvoker = rootMirror.getClassIfDefined(invokerFullName)
-        if (!currentRun.compiles(existingInvoker)) {
+        // synthesize the bundle, i.e. given a top-level `trait Foo extends Macro { def expand = ... } `
+        // create a top-level definition `class FooBundle(val c: Context) extends Foo` in MACRO_BUNDLE_PACKAGE
+        val bundlePid = gen.mkUnattributedRef(nme.MACRO_BUNDLE_PACKAGE)
+        val bundleName = TypeName(bundleProto.fullName.split('.').map(_.capitalize).mkString("Dot") + nme.MACRO_BUNDLE_SUFFIX)
```
Comment (retronym) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: Okay, I see you use fully qualified names here. It would be more standard to encode them as `com$acme$Foo$Bundle` in the same way as the compiler does in `expandName` rather than inventing a new means to do this.

In any case, I'm still against putting them in this package, but I'm keen to hear what constraints led to this design.

(Sorry that I'm re-reviewing this old code now, I must have missed it when it went on originally.)

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: The first line should say "static", the second line should remain "top-level".

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala (hunk 2 above)
Comment (xeno-by) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: You're right

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala (hunk 3 above)
Comment (xeno-by) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: Thanks

File: src/reflect/scala/reflect/internal/StdNames.scala (hunk 4 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/StdNames.scala]: Done

File: src/reflect/scala/reflect/internal/StdNames.scala (hunk 5 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/StdNames.scala]: You're right. I'll move synthetic code to the user's package.

File: src/compiler/scala/reflect/macros/compiler/Resolvers.scala (hunk 6 above)
Comment (xeno-by) [inline src/compiler/scala/reflect/macros/compiler/Resolvers.scala]: Okay

**Comment by xeno-by** [comment]
@retronym All feedback is addressed, and an SBT blocker is now fixed. Would be great to get a go for M6.

**Comment by xeno-by** [comment]
Rebased and split into more clear commits.

**Comment by retronym** [comment]
LGTM. Thanks for addressing all of my feedback so quickly!
